File: 300/304 - Range Sum Query 2D - Immutable.java
public class NumMatrix {
    
    int dp[][];

    public NumMatrix(int[][] matrix) {
        int i, j;
        
        this.dp = matrix;
        
        if (matrix.length == 0 || matrix[0].length == 0)
            return;
        
        for (i = 1; i < matrix.length; ++i)
            this.dp[i][0] += this.dp[i - 1][0];
        
        for (j = 1; j < matrix[0].length; ++j)
            this.dp[0][j] += this.dp[0][j - 1];
        
        for (i = 1; i < matrix.length; ++i) {
            for (j = 1; j < matrix[0].length; ++j)
                this.dp[i][j] += this.dp[i - 1][j] + this.dp[i][j - 1] - this.dp[i - 1][j - 1];
        }
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        if (row1 == 0 && col1 == 0)
            return dp[row2][col2];
        else if (row1 == 0)
            return dp[row2][col2] - dp[row2][col1 - 1];  
        else if (col1 == 0)
            return dp[row2][col2] - dp[row1 - 1][col2];
        return dp[row2][col2] + dp[row1 - 1][col1 - 1] - dp[row1 - 1][col2] - dp[row2][col1 - 1];
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */
 

File: csharp/304.range-sum-query-2-d-immutable.cs
/*
 * @lc app=leetcode id=304 lang=csharp
 *
 * [304] Range Sum Query 2D - Immutable
 */
public class NumMatrix {

    private int[,] sum = null;

    public NumMatrix(int[][] matrix) {
        int m = matrix.Length;
        if (m == 0) {
            return;
        }
        int n = matrix[0].Length;
        sum = new int[m + 1, n + 1];
        sum[1, 1] = matrix[0][0];
        for (int j = 1; j < n; ++j) {
            sum[1, j + 1] = matrix[0][j] + sum[1, j];
        }
        for (int i = 1; i < m; ++i) {
            sum[i + 1, 1] = matrix[i][0] + sum[i, 1];
            for (int j = 1; j < n; ++j) {
                sum[i + 1, j + 1] = sum[i, j + 1] + sum[i + 1, j] - sum[i, j] + matrix[i][j];
            }
        }
    }
    
    public int SumRegion(int row1, int col1, int row2, int col2) {
        return sum == null ? 0 : (sum[row2 + 1, col2 + 1] + sum[row1, col1] - sum[row1, col2 + 1] - sum[row2 + 1, col1]);
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.SumRegion(row1,col1,row2,col2);
 */

